### PR TITLE
Refactor Stripe SCA backoffice payment authorization

### PR DESCRIPTION
**Pull request by sauloperez**
#### What? Why?

This removes the tight coupling of the /admin/payments_controller, the one dealing with backoffice payments, with StripeSCA. It's a refactoring that will ease future changes. I won't get into details about why coupling is bad. That's well known, I think.

Here are the things we did to solve this:

* Replace the offending line `return unless @payment.payment_method.instance_of?(Spree::Gateway::StripeSCA)` with polymorphism. Spree's Gateways are meant to stick to a shared interface between all them, implementing the [Adapter pattern](https://en.wikipedia.org/wiki/Adapter_pattern) I'd say, so that consumers are agnostic of the particular details each gateway is concerned with. In this case, most of them don't require authorization. That's why we implemented this operation as a no op :ok_hand: 
* Decoupling the controller from the payment states by checking state machine errors. If `Payment::Processing` fails to change the payment state, it'll let us know with a `StateMachine::InvalidTransition`. No need to check for each possible state as done in https://github.com/openfoodfoundation/openfoodnetwork/pull/7809/files#diff-e3391ddb14a78e0ed8ced3c32cd647c691a34a0e18590e64d4a262e481d66fceL179-L181. What if they change as happened recently? this code would break.
* Replace exceptions as a way to reuse logic. That's what https://github.com/openfoodfoundation/openfoodnetwork/pull/7809/files#diff-e3391ddb14a78e0ed8ced3c32cd647c691a34a0e18590e64d4a262e481d66fceL185-L186 were doing. That's a confusing way to do that because it's not an error scenario.
* Extend the controller's test coverage. There was a case that wasn't covered. This makes it safe to deal with future refactorings.

Now, this is a good middle grown. The situation is better than the one we found, it's easy to review and revert if needed. Going any further (I did) clutters the diff too much. What @andrewpbrett and I agree needs to be done now is to extract that logic out in a service (or any other abstractions) so that we leave the `flash` and `redirect_to` calls in the controller but the business logic is nicely encapsulated elsewhere. That will remove the duplication we left there.

#### What should we test?

backoffice payments with the supported gateways: Cash (no action?), PayPal, StripeSCA, and Stripe Connect can probably be skipped. Although we do have implementations for Bogus and Bogus Simple and not entirely sure we use them. Time for a DB query from Metabase to get some counts.

Now, if you catch any regression @filipefurtad0 please, let's create an issue so we can keep track of it. I did pay attention to controller tests but I haven't checked feature ones.

#### Release notes
Refactor Stripe SCA backoffice payment authorization to lower the coupling in the controller
Changelog Category: Technical changes

### CODE REVIEW

File: spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb [hunk 1]
```diff
@@ -40,7 +40,7 @@
           spree_post :create, payment: params, order_id: order.number
 
           redirects_to_list_of_payments_with_success_flash
-          expect(order.reload.payments.last.state).to eq "checkout"
+          expect(order.reload.payments.last.state).to eq "pending"
```
Comment (sauloperez) [inline spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb:43]: This is a meaningful change

File: app/models/spree/payment/processing.rb [hunk 2]
```diff
@@ -23,6 +23,7 @@ def process_offline!
       def authorize!(return_url = nil)
         started_processing!
         gateway_action(source, :authorize, :pend, return_url: return_url)
+        raise Spree::Core::GatewayError, I18n.t('authorization_failure') unless pending?
```
Comment (sauloperez) [inline app/models/spree/payment/processing.rb:26]: that's something that only applies to StripeSCA, so we should move it deeper into `Gateway::StripeSCA`.

File: app/controllers/spree/admin/payments_controller.rb [hunk 3]
```diff
@@ -171,12 +171,8 @@ def load_payment
       end
 
       def authorize_stripe_sca_payment
-        return unless @payment.payment_method.class == Spree::Gateway::StripeSCA
```
Comment (sauloperez) [inline app/controllers/spree/admin/payments_controller.rb:174]: it turns out backoffice payments only allow Cash and StripeSCA, so we better validate that as early as possible. Regardless of the authorization, etc.

File: spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb (hunk 1 above)
Comment (sauloperez) [reply spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb:43]: This may be confusing for users that use it a lot because now order payments will display this state once placed. Something to discuss with product/Filipe.

File: app/models/spree/payment/processing.rb (hunk 2 above)
Comment (sauloperez) [reply app/models/spree/payment/processing.rb:26]: If we can't progress to `pending` it might be because a model validation failed. In that case, we might want to see that in Bugsnag, and it has nothing to do with Gateway errors. It feels like we're missing concerns.

File: app/controllers/spree/admin/payments_controller.rb (hunk 3 above)
Comment (sauloperez) [reply app/controllers/spree/admin/payments_controller.rb:174]: It can show up if the enterprise has it configured but nothing is stopping you from using it :shrug: 

Screenshot from 2021-06-21 11-47-52
Screenshot from 2021-06-21 11-48-20

IMO this needs to be reconsider but it falls out of the scope of this PR.

File: app/models/spree/payment/processing.rb (hunk 2 above)
Comment (sauloperez) [reply app/models/spree/payment/processing.rb]: @andrewpbrett as discussed, a7e1b7e67bde000fe3beb0941e17a051bc576e3a makes things clear while keeping the behaviour.

File: app/models/spree/payment/processing.rb [hunk 4]
```diff
@@ -25,6 +25,13 @@ def authorize!(return_url = nil)
         gateway_action(source, :authorize, :pend, return_url: return_url)
       end
 
+      def request_user_authorization
+        return unless authorization_action_required?
```
Comment (sauloperez) [inline app/models/spree/payment/processing.rb:29]: we can move this deeper, and make each gateway determine if there's further action required.

File: spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb (hunk 1 above)
Comment (sauloperez) [reply spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb:43]: :warning: this could clash with https://github.com/openfoodfoundation/openfoodnetwork/blob/7b87476986aeb0db64624bdb661fcf4e63e37357/app/services/order_syncer.rb#L58-L69. It checks for checkout state.

Comment (sauloperez) [reply spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb:55]: let's see if what I found out makes sense. This code above checks for unprocessed payments placed with the subs' payment method that was just changed (these are the sort of things Order Syncer syncs). Since the scope of the present PR is backoffice payments only, there are no consequences that could affect subscriptions.

While investigating I found out we check for `checkout` because that's the initial state of the payment's state machine. Yes, this code could do better at being a bit more intention revealing and removing duplication. It's the [SubscriptionConfirmJob#process_payment!](https://github.com/openfoodfoundation/openfoodnetwork/blob/b1324c62698867f9c1870b302ec579be146a72c1/app/jobs/subscription_confirm_job.rb#L52) which moves the payment through the state machine.

So, tl;dr. I think we're safe here @andrewpbrett

**Comment by sauloperez** [comment]
I had to fix the conflicts caused by https://github.com/openfoodfoundation/openfoodnetwork/pull/7781/files#diff-e3391ddb14a78e0ed8ced3c32cd647c691a34a0e18590e64d4a262e481d66fce. I'd appreciate it if you could pay extra attention at that while code reviewing.

**Comment by RachL** [comment]
@Matt-Yorkley same question here: is this PR still interesting after your refactoring?

**Comment by Matt-Yorkley** [comment]
Hmmm... it's tricky. There's some interesting stuff here, but the PR is ~8 months old and we've changed various things in that time so there's lots of conflicts. It might take a fair bit of work to bring it up to date.

**Comment by RachL** [comment]
OK then I'm closing it. If anyone disagree feel free to reopen.
